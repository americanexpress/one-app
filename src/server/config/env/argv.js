/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import yargs, { argv } from 'yargs';
import path from 'path';

const rootModuleNameEnvVarValue = process.env.ONE_CLIENT_ROOT_MODULE_NAME;

if (process.env.NODE_ENV === 'development') {
  yargs
    .option('m', {
      alias: 'use-middleware',
      describe: 'Apply a custom middleware configuration for one-app-dev-proxy',
      type: 'boolean',
    })
    .option('root-module-name', {
      describe: 'Name of the module that serves as the entry point to your application.',
      type: 'string',
    })
    .option('module-map-url', {
      describe: 'Remote module map URL for one-app-dev-cdn to proxy',
      type: 'string',
    })
    .option('use-host', {
      describe: 'use req.headers.host instead of localhost for one-app-dev-cdn',
      default: false,
      type: 'boolean',
    });

  // only require a remote module map if there are no locally served modules
  let isLocalModuleMapEmpty = false;
  try {
    // eslint-disable-next-line import/no-dynamic-require,global-require
    const localModuleMap = require(path.join(process.cwd(), 'static', 'module-map.json'));
    if (Object.entries(localModuleMap).length === 0) {
      isLocalModuleMapEmpty = true;
    }
  // if local module map is an empty file we need to catch it
  } catch (error) {
    isLocalModuleMapEmpty = true;
  }

  if (isLocalModuleMapEmpty) {
    if (!rootModuleNameEnvVarValue) {
      yargs.demandOption(
        ['module-map-url', 'root-module-name'],
        'Both `root-module-name` and `module-map-url` args are required if there are no locally served modules.'
      );
    } else {
      yargs.demandOption(
        ['module-map-url'],
        '`module-map-url` is required if there are no locally served modules.'
      );
    }
  } else if (!rootModuleNameEnvVarValue) {
    yargs.demandOption(
      ['root-module-name'],
      '`root-module-name` is a required arg.'
    );
  }
}

yargs
  .option('log-format', {
    describe: 'Presentation of log entries',
    type: 'string',
    choices: ['friendly', 'verbose', 'machine'],
    default: process.env.NODE_ENV === 'development' ? 'friendly' : 'machine',
  })
  .option('log-level', {
    describe: 'Lowest level of log entries to show',
    type: 'string',
    choices: ['error', 'warn', 'log', 'info'],
    default: process.env.NODE_ENV === 'development' ? 'log' : 'info',
  });

if (argv.rootModuleName && rootModuleNameEnvVarValue) {
  throw new Error(
    'Both the `root-module-name` argument and the `ONE_CLIENT_ROOT_MODULE_NAME` environment variable have been provided, but only one may be set at once.\n'
  );
}
