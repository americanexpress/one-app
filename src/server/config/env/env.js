/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { argv } from 'yargs';
import path from 'path';

function getLocalModuleMap() {
  const localModulePath = path.join(process.cwd(), 'static', 'module-map.json');
  try {
    // eslint-disable-next-line import/no-dynamic-require,global-require
    return require(localModulePath);
  } catch (e) {
    return 'no such path';
  }
}

const moduleMap = argv.moduleMapUrl || getLocalModuleMap();
const rootModuleName = process.env.ONE_CLIENT_ROOT_MODULE_NAME || argv.rootModuleName;

const environment = {
  NODE_ENV: process.env.NODE_ENV,
  rootModuleName,
  HOLOCRON_MODULE_MAP_URL: process.env.HOLOCRON_MODULE_MAP_URL,
  ONE_CLIENT_REPORTING_URL: process.env.ONE_CLIENT_REPORTING_URL,
  ONE_CLIENT_CSP_REPORTING_URL: process.env.ONE_CLIENT_CSP_REPORTING_URL,
  ONE_CLIENT_ROOT_MODULE_NAME: process.env.ONE_CLIENT_ROOT_MODULE_NAME,
  ONE_CLIENT_CDN_URL: process.env.ONE_CLIENT_CDN_URL,
  ONE_CONFIG_ENV: process.env.ONE_CONFIG_ENV,
  moduleMap,
  rootModuleNameDuplicate: Boolean(argv.rootModuleName && process.env.ONE_CLIENT_ROOT_MODULE_NAME),
};

export { environment as default };
