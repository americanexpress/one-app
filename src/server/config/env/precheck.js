import { argv } from 'yargs';
import isFetchableUrlInNode from '@americanexpress/env-config-utils/isFetchableUrlInNode';
import isFetchableUrlInBrowser from '@americanexpress/env-config-utils/isFetchableUrlInBrowser';
import path from 'path';

// The required environment variables and config options.
const environment = {
  NODE_ENV: process.env.NODE_ENV,
  rootModuleName: argv.rootModuleName,
  HOLOCRON_MODULE_MAP_URL: process.env.HOLOCRON_MODULE_MAP_URL,
  ONE_CLIENT_REPORTING_URL: process.env.ONE_CLIENT_REPORTING_URL,
  ONE_CLIENT_CSP_REPORTING_URL: process.env.ONE_CLIENT_CSP_REPORTING_URL,
  ONE_CLIENT_ROOT_MODULE_NAME: process.env.ONE_CLIENT_ROOT_MODULE_NAME,
  ONE_CLIENT_CDN_URL: process.env.ONE_CLIENT_CDN_URL,
  ONE_CONFIG_ENV: process.env.ONE_CONFIG_ENV,
  localModuleMap: argv.localModuleMap,
  rootModuleNameDuplicate: Boolean(argv.rootModuleName && process.env.ONE_CLIENT_ROOT_MODULE_NAME),
};

// The rules the environment must follow
const rules = {
  NODE_ENV(parameter) {
    return {
      success: ['development', 'production']
        .some((option) => option === parameter),
      explanation: `NODE_ENV environment variable must be either "development" or "production"; given: ${parameter}`,
    };
  },
  rootModuleName(parameter) {
    return {
      success: Boolean(parameter) === true,
      explanation: 'option --root-module-name=<module-name> must be defined if NODE_ENV != "development"',
    };
  },
  HOLOCRON_MODULE_MAP_URL(parameter) {
    let success = true;
    let message;
    try {
      isFetchableUrlInNode(parameter);
    } catch (e) {
      success = false;
      message = e.message;
    }
    return {
      success,
      explanation: `HOLOCRON_MODULE_MAP_URL: ${message}`,
    };
  },
  ONE_CLIENT_REPORTING_URL(parameter) {
    let success = true;
    let message;
    try {
      isFetchableUrlInBrowser(parameter);
    } catch (e) {
      success = false;
      message = e.message;
    }
    return {
      success,
      explanation: `ONE_CLIENT_REPORTING_URL: ${message}`,
    };
  },
  ONE_CLIENT_CSP_REPORTING_URL(parameter) {
    let success = true;
    let message;
    try {
      isFetchableUrlInBrowser(parameter);
    } catch (e) {
      success = false;
      message = e.message;
    }
    return {
      success,
      explanation: `ONE_CLIENT_CSP_REPORTING_URL: ${message}`,
    };
  },
  ONE_CLIENT_ROOT_MODULE_NAME(parameter) {
    return {
      success: Boolean(parameter) === true,
      explanation: 'ONE_CLIENT_ROOT_MODULE_NAME environment variable must be defined',
    };
  },
  ONE_CLIENT_CDN_URL(parameter) {
    let success = true;
    let message;
    try {
      isFetchableUrlInBrowser(parameter);
      if (!/\/$/.test(parameter)) {
        throw new Error('ONE_CDN_URL must have a trailing slash');
      }
    } catch (e) {
      success = false;
      message = e.message;
    }
    return {
      success,
      explanation: `ONE_CDN_URL: ${message}`,
    };
  },
  ONE_CONFIG_ENV(parameter) {
    return {
      success: Boolean(parameter) === true,
      explanation: 'ONE_CONFIG_ENV environment variable must be defined',
    };
  },
  localModuleMap(parameter) {
    // only require a remote module map if there are no locally served modules
    let isLocalModuleMapEmpty = false;
    try {
      // eslint-disable-next-line import/no-dynamic-require,global-require
      const localModuleMap = require(path.join(process.cwd(), 'static', 'module-map.json'));
      if (Object.entries(localModuleMap).length === 0) {
        isLocalModuleMapEmpty = true;
      }
      // if local module map is an empty file we need to catch it
    } catch (error) {
      isLocalModuleMapEmpty = true;
    }
    return {
      success: !isLocalModuleMapEmpty || Boolean(parameter) === true,
      explanation: '`module-map-url` is required if there are no locally served modules',
    };
  },
  rootModuleNameDuplicate(parameter) {
    return {
      success: !parameter,
      explanation: 'Both the `root-module-name` argument and the `ONE_CLIENT_ROOT_MODULE_NAME` environment variable have been provided, but only one may be set at once.\n',
    };
  },
};


function selectRules() {
  const {
    NODE_ENV,
    rootModuleName,
    HOLOCRON_MODULE_MAP_URL,
    ONE_CLIENT_REPORTING_URL,
    ONE_CLIENT_CSP_REPORTING_URL,
    ONE_CLIENT_ROOT_MODULE_NAME,
    ONE_CLIENT_CDN_URL,
    ONE_CONFIG_ENV,
    localModuleMap,
    rootModuleNameDuplicate,
  } = rules;
  if (environment.NODE_ENV === 'development') {
    return {
      rootModuleName,
      localModuleMap,
    };
  }
  if (environment.NODE_ENV === 'production') {
    return {
      HOLOCRON_MODULE_MAP_URL,
      ONE_CLIENT_REPORTING_URL,
      ONE_CLIENT_CSP_REPORTING_URL,
      ONE_CLIENT_ROOT_MODULE_NAME,
      ONE_CLIENT_CDN_URL,
      ONE_CONFIG_ENV,
      rootModuleNameDuplicate,
    };
  }
  return {
    NODE_ENV,
  };
}

// An object mapping rule names to explanations for why they were broken
const rulesBroken = Object.entries(selectRules())
  .reduce((results, [parameter, validator]) => {
    const { success, explanation } = validator(environment[parameter]);
    if (!success) {
      results[parameter] = explanation;
    }
    return results;
  }, {});

if (Object.keys(rulesBroken).length > 0) {
  console.error('ERRORS:', rulesBroken);
  throw new Error('Please fix the above errors');
}
