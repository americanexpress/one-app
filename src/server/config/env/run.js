import environment from './env';
import validateEnvironment from './precheck';

validateEnvironment(environment);
