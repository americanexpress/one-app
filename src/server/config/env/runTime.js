/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

const ip = require('ip').address();
const { preprocessEnvVar } = require('@americanexpress/env-config-utils');
const isFetchableUrlInNode = require('@americanexpress/env-config-utils/isFetchableUrlInNode');
const isFetchableUrlInBrowser = require('@americanexpress/env-config-utils/isFetchableUrlInBrowser');
const { argv } = require('yargs');

const isPositiveIntegerIfDefined = (input) => {
  if (input === undefined) {
    return;
  }

  const value = Number(input);

  if (value > 0 && value % 1 === 0) {
    return;
  }

  throw new Error(`Expected ${input} to be a positive integer`);
};

// the set of env vars to validate and normalize
// order matters, ex: ONE_CLIENT_REPORTING_URL uses HTTP_PORT
const runTime = [
  {
    name: 'NODE_ENV',
    normalize: (input) => input && input.toLowerCase(),
    valid: ['development', 'production'],
    defaultValue: 'production',
  },
  // heapdump config
  {
    name: 'NODE_HEAPDUMP_OPTIONS',
    defaultValue: 'nosignal',
  },
  // IPv4 port to bind on
  {
    name: 'HTTP_PORT',
    normalize: (input) => {
      const parsed = parseInt(input, 10);
      // make sure the parsed value is the same value as input
      // input may be a string or a number, we don't want === in this case, just ==
      if (Number.isNaN(parsed) || parsed != input) { // eslint-disable-line eqeqeq
        throw new Error(`env var HTTP_PORT needs to be a valid integer, given "${input}"`);
      } else {
        return parsed;
      }
    },
    defaultValue: () => {
      if (process.env.PORT) { return process.env.PORT; }
      return 3000;
    },
  },
  // IPv4 port for the health and metrics server to bind on
  {
    name: 'HTTP_METRICS_PORT',
    normalize: (input) => {
      const parsed = parseInt(input, 10);
      // make sure the parsed value is the same value as input
      // input may be a string or a number, we don't want === in this case, just ==
      if (Number.isNaN(parsed) || parsed != input) { // eslint-disable-line eqeqeq
        throw new Error(`env var HTTP_METRICS_PORT needs to be a valid integer, given "${input}"`);
      } else {
        return parsed;
      }
    },
    defaultValue: () => 3005,
  },
  {
    name: 'HTTP_ONE_APP_DEV_CDN_PORT',
    normalize: (input) => {
      if (input) {
        const parsed = parseInt(input, 10);
        // make sure the parsed value is the same value as input
        // input may be a string or a number, we don't want === in this case, just ==
        if (Number.isNaN(parsed) || parsed != input) { // eslint-disable-line eqeqeq
          throw new Error(`env var HTTP_ONE_APP_DEV_CDN_PORT needs to be a valid integer, given "${input}"`);
        } else {
          return parsed;
        }
      }
      return undefined;
    },
    defaultValue: () => (process.env.NODE_ENV === 'development'
      ? 3001
      : undefined),
  },
  {
    name: 'HTTP_ONE_APP_DEV_PROXY_SERVER_PORT',
    normalize: (input) => {
      if (input) {
        const parsed = parseInt(input, 10);
        // make sure the parsed value is the same value as input
        // input may be a string or a number, we don't want === in this case, just ==
        if (Number.isNaN(parsed) || parsed != input) { // eslint-disable-line eqeqeq
          throw new Error(`env var HTTP_ONE_APP_DEV_PROXY_SERVER_PORT needs to be a valid integer, given "${input}"`);
        } else {
          return parsed;
        }
      }
      return undefined;
    },
    defaultValue: () => (process.env.NODE_ENV === 'development'
      ? 3002
      : undefined),
  },
  // holocron config, the modules to use
  {
    name: 'HOLOCRON_MODULE_MAP_URL',
    defaultValue: () => (process.env.NODE_ENV === 'development'
      ? `http://${ip}:${process.env.HTTP_ONE_APP_DEV_CDN_PORT}/static/module-map.json`
      : undefined),
    validate: isFetchableUrlInNode,
  },
  {
    name: 'HOLOCRON_SERVER_MAX_MODULES_RETRY',
    validate: isPositiveIntegerIfDefined,
  },
  {
    name: 'HOLOCRON_SERVER_MAX_SIM_MODULES_FETCH',
    validate: isPositiveIntegerIfDefined,
  },
  // where to send/report client errors
  {
    name: 'ONE_CLIENT_REPORTING_URL',
    defaultValue: () => (process.env.NODE_ENV === 'development'
      ? `http://${ip}:${process.env.HTTP_PORT}/_`
      : undefined),
    validate: isFetchableUrlInBrowser,
  },
  // allow hosting of static assets externally
  {
    name: 'ONE_CLIENT_CDN_URL',
    defaultValue: () => (process.env.NODE_ENV === 'development'
      ? '/_/static/'
      : undefined),
    validate: (value) => {
      isFetchableUrlInBrowser(value);
      if (!/\/$/.test(value)) {
        throw new Error('ONE_CDN_URL must have a trailing slash');
      }
    },
  },
  // locale folder env level
  {
    name: 'ONE_CLIENT_LOCALE_FILENAME',
    valid: ['integration', 'qa', undefined],
    normalize: (input) => input || undefined,
    defaultValue: () => (process.env.NODE_ENV === 'development' ? 'integration' : undefined),
  },
  {
    name: 'ONE_CLIENT_ROOT_MODULE_NAME',
    validate: (value) => { if (!value) { throw new Error('The `ONE_CLIENT_ROOT_MODULE_NAME` environment variable must be defined.'); } },
    defaultValue: () => (process.env.NODE_ENV === 'development' ? argv.rootModuleName : undefined),
  },
];

runTime.forEach(preprocessEnvVar);
export { ip };
export default runTime;
