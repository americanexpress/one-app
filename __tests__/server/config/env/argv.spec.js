/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import fs from 'fs';
import path from 'path';

const defaultExport = {
  option: jest.fn(() => defaultExport),
  demandOption: jest.fn(() => defaultExport),
};

const yargsMock = {
  __esModule: true,
  default: defaultExport,
  argv: { rootModuleName: 'frank-lloyd-root' },
};
jest.doMock('yargs', () => yargsMock);

const load = () => {
  jest.resetModules();
  require('../../../../src/server/config/env/argv');
};

const pathToModuleMap = path.join(process.cwd(), 'static', 'module-map.json');

let origModuleMapContents;

beforeAll(() => {
  if (fs.existsSync(pathToModuleMap)) {
    origModuleMapContents = fs.readFileSync(pathToModuleMap, 'utf8');
    fs.unlinkSync(pathToModuleMap);
  }
});

afterAll(() => {
  if (origModuleMapContents) {
    fs.writeFileSync(pathToModuleMap, origModuleMapContents, 'utf8');
  }
});

describe('argv config', () => {
  it('has production options and defaults', () => {
    process.env.NODE_ENV = 'production';
    yargsMock.default.option.mockClear();
    yargsMock.default.demandOption.mockClear();
    load();
    expect(yargsMock.default.option).toHaveBeenCalled();
    yargsMock.default.option.mock.calls.forEach((args) => {
      expect(args).toMatchSnapshot();
    });
  });

  it('has development options and defaults', () => {
    process.env.NODE_ENV = 'development';
    yargsMock.default.option.mockClear();
    yargsMock.default.demandOption.mockClear();
    jest.doMock(pathToModuleMap, () => ({ 'my-module': '1.0.0' }), { virtual: true });
    load();
    expect(yargsMock.default.option).toHaveBeenCalled();
    expect(yargsMock.default.demandOption).toHaveBeenCalled();
    yargsMock.default.option.mock.calls.forEach((args) => {
      expect(args).toMatchSnapshot();
    });
    expect(yargsMock.default.demandOption.mock.calls).toMatchSnapshot();
  });

  it('requires the `module-map-url` option if there are no locally served modules', () => {
    process.env.NODE_ENV = 'development';
    yargsMock.default.option.mockClear();
    yargsMock.default.demandOption.mockClear();
    jest.dontMock(pathToModuleMap);
    load();
    expect(yargsMock.default.demandOption).toHaveBeenCalled();
    expect(yargsMock.default.demandOption.mock.calls).toMatchSnapshot();
  });

  it('requires the `module-map-url` option if there is a local module map that is empty', () => {
    process.env.NODE_ENV = 'development';
    yargsMock.default.option.mockClear();
    yargsMock.default.demandOption.mockClear();
    jest.doMock(pathToModuleMap, () => ({}), { virtual: true });
    load();
    expect(yargsMock.default.demandOption).toHaveBeenCalled();
    expect(yargsMock.default.demandOption.mock.calls).toMatchSnapshot();
  });

  it('requires the `module-map-url` and `root-module-name` options if there is a local module map that is empty and the root module name env var is not set', () => {
    process.env.NODE_ENV = 'development';
    delete process.env.ONE_CLIENT_ROOT_MODULE_NAME;

    yargsMock.default.option.mockClear();
    yargsMock.default.demandOption.mockClear();
    jest.doMock(pathToModuleMap, () => ({}), { virtual: true });
    load();
    expect(yargsMock.default.demandOption).toHaveBeenCalled();
    expect(yargsMock.default.demandOption.mock.calls).toMatchSnapshot();
  });

  it('requires the `module-map-url` but not the `root-module-name` option if there is a local module map that is empty and the root module name env var is set', () => {
    process.env.NODE_ENV = 'development';
    process.env.ONE_CLIENT_ROOT_MODULE_NAME = 'frank-lloyd-root';
    yargsMock.argv = {};
    yargsMock.default.option.mockClear();
    yargsMock.default.demandOption.mockClear();
    jest.doMock(pathToModuleMap, () => ({}), { virtual: true });
    load();
    expect(yargsMock.default.demandOption).toHaveBeenCalled();
    expect(yargsMock.default.demandOption.mock.calls).toMatchSnapshot();
  });

  it('does not require either the `module-map-url` or the `root-module-name` option if there is a local module map and the root module name env var is set', () => {
    process.env.NODE_ENV = 'development';
    process.env.ONE_CLIENT_ROOT_MODULE_NAME = 'frank-lloyd-root';
    yargsMock.argv = {};
    yargsMock.default.option.mockClear();
    yargsMock.default.demandOption.mockClear();
    jest.doMock(pathToModuleMap, () => ({ 'my-module': '1.0.0' }), { virtual: true });
    load();
    expect(yargsMock.default.demandOption).not.toHaveBeenCalled();
  });

  it('throws if the `root-module-name` option is given while the root module name env var is also set', () => {
    process.env.NODE_ENV = 'development';
    process.env.ONE_CLIENT_ROOT_MODULE_NAME = 'frank-lloyd-root';
    yargsMock.argv = { rootModuleName: 'frank-lloyd-root' };
    yargsMock.default.option.mockClear();
    yargsMock.default.demandOption.mockClear();
    jest.doMock(pathToModuleMap, () => ({}), { virtual: true });
    expect(() => load()).toThrowErrorMatchingSnapshot();
  });
});
