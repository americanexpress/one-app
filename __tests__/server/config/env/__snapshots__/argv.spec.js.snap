// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`argv config has development options and defaults 1`] = `
Array [
  "m",
  Object {
    "alias": "use-middleware",
    "describe": "Apply a custom middleware configuration for one-app-dev-proxy",
    "type": "boolean",
  },
]
`;

exports[`argv config has development options and defaults 2`] = `
Array [
  "root-module-name",
  Object {
    "describe": "Name of the module that serves as the entry point to your application.",
    "type": "string",
  },
]
`;

exports[`argv config has development options and defaults 3`] = `
Array [
  "module-map-url",
  Object {
    "describe": "Remote module map URL for one-app-dev-cdn to proxy",
    "type": "string",
  },
]
`;

exports[`argv config has development options and defaults 4`] = `
Array [
  "use-host",
  Object {
    "default": false,
    "describe": "use req.headers.host instead of localhost for one-app-dev-cdn",
    "type": "boolean",
  },
]
`;

exports[`argv config has development options and defaults 5`] = `
Array [
  "log-format",
  Object {
    "choices": Array [
      "friendly",
      "verbose",
      "machine",
    ],
    "default": "friendly",
    "describe": "Presentation of log entries",
    "type": "string",
  },
]
`;

exports[`argv config has development options and defaults 6`] = `
Array [
  "log-level",
  Object {
    "choices": Array [
      "error",
      "warn",
      "log",
      "info",
    ],
    "default": "log",
    "describe": "Lowest level of log entries to show",
    "type": "string",
  },
]
`;

exports[`argv config has development options and defaults 7`] = `
Array [
  Array [
    Array [
      "root-module-name",
    ],
    "\`root-module-name\` is a required arg.",
  ],
]
`;

exports[`argv config has production options and defaults 1`] = `
Array [
  "log-format",
  Object {
    "choices": Array [
      "friendly",
      "verbose",
      "machine",
    ],
    "default": "machine",
    "describe": "Presentation of log entries",
    "type": "string",
  },
]
`;

exports[`argv config has production options and defaults 2`] = `
Array [
  "log-level",
  Object {
    "choices": Array [
      "error",
      "warn",
      "log",
      "info",
    ],
    "default": "info",
    "describe": "Lowest level of log entries to show",
    "type": "string",
  },
]
`;

exports[`argv config requires the \`module-map-url\` and \`root-module-name\` options if there is a local module map that is empty and the root module name env var is not set 1`] = `
Array [
  Array [
    Array [
      "module-map-url",
      "root-module-name",
    ],
    "Both \`root-module-name\` and \`module-map-url\` args are required if there are no locally served modules.",
  ],
]
`;

exports[`argv config requires the \`module-map-url\` but not the \`root-module-name\` option if there is a local module map that is empty and the root module name env var is set 1`] = `
Array [
  Array [
    Array [
      "module-map-url",
    ],
    "\`module-map-url\` is required if there are no locally served modules.",
  ],
]
`;

exports[`argv config requires the \`module-map-url\` option if there are no locally served modules 1`] = `
Array [
  Array [
    Array [
      "module-map-url",
      "root-module-name",
    ],
    "Both \`root-module-name\` and \`module-map-url\` args are required if there are no locally served modules.",
  ],
]
`;

exports[`argv config requires the \`module-map-url\` option if there is a local module map that is empty 1`] = `
Array [
  Array [
    Array [
      "module-map-url",
      "root-module-name",
    ],
    "Both \`root-module-name\` and \`module-map-url\` args are required if there are no locally served modules.",
  ],
]
`;

exports[`argv config throws if the \`root-module-name\` option is given while the root module name env var is also set 1`] = `
"Both the \`root-module-name\` argument and the \`ONE_CLIENT_ROOT_MODULE_NAME\` environment variable have been provided, but only one may be set at once.
"
`;
